File: plug-ins/dev/python/n_gfUtilEulerScalarMath.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

MIT License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

====================================================================================================

Disclaimer:
    THIS PLUGIN IS JUST A PROTOTYPE. YOU MUST USE THE C++ RELEASE PLUGIN FOR PRODUCTION.
    YOU CAN FIND THE C++ RELEASE PLUGIN FOR YOUR SPECIFIC PLATFORM IN RELEASES FOLDER:
    "gfTools > plug-ins > release"

How to use:
    * Copy the parent folder to the MAYA_SCRIPT_PATH.
    * To find MAYA_SCRIPT_PATH paste this command in a Python tab:
        import os; os.environ["MAYA_SCRIPT_PATH"].split(";")
    * In Maya, go to Windows > Settings/Preferences > Plug-in Manager.
    * Browse for "gfTools > plug-ins > dev > python"
    * Find gfTools_P.py and import it.

Requirements:
    * Maya 2017 or above.

Description:
    Basic euler rotation operations with float scalar. No operation value returns the euler rotation.
    Support addition, subtraction and multiply.

Attributes:
    * Operation: The math operation. Can be add, subtract or multiply.
    * Euler: The euler rotation of the operation.
    * Euler Rotate Order: The rotation order of the euler rotation.
    * Scalar: The scalar value of the operation.
    * Rotate Order Out Euler: The rotation order of the output euler rotation.
    * Out Euler: The output euler rotation.

Todo:
    * NDA

Sources:
    * NDA

This code supports Pylint. Rc file in project.
"""
import maya.api._OpenMaya_py2 as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=invalid-name, unnecessary-pass
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class EulerScalarMath(om2.MPxNode):
    """ Main class of gfUtilEulerScalarMath node. """

    kNodeName = ""
    kNodeClassify = ""
    kNodeID = ""

    inOperation = om2.MObject()
    inEuler = om2.MObject()
    inEulerRotOrder = om2.MObject()
    inScalar = om2.MObject()
    inResRotOrder = om2.MObject()
    outEuler = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return EulerScalarMath()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to EulerScalarMath class. Instances of EulerScalarMath will use these attributes to create plugs
        for use in the compute() method.
        """
        eAttr = om2.MFnEnumAttribute()
        uAttr = om2.MFnUnitAttribute()
        nAttr = om2.MFnNumericAttribute()

        EulerScalarMath.inOperation = eAttr.create("operation", "operation", 0)
        eAttr.addField("No Operation", 0)
        eAttr.addField("Add", 1)
        eAttr.addField("Subtract", 2)
        eAttr.addField("Multiply", 3)
        INPUT_ATTR(eAttr)

        eulerX = uAttr.create("eulerX", "ex", om2.MFnUnitAttribute.kAngle, 0.0)
        eulerY = uAttr.create("eulerY", "ey", om2.MFnUnitAttribute.kAngle, 0.0)
        eulerZ = uAttr.create("eulerZ", "ez", om2.MFnUnitAttribute.kAngle, 0.0)
        EulerScalarMath.inEuler = nAttr.create("euler", "e", eulerX, eulerY, eulerZ)
        INPUT_ATTR(nAttr)

        EulerScalarMath.inEulerRotOrder = eAttr.create("rotateOrderEuler", "roe", 0)
        eAttr.addField("xyz", 0)
        eAttr.addField("yzx", 1)
        eAttr.addField("zxy", 2)
        eAttr.addField("xzy", 3)
        eAttr.addField("yxz", 4)
        eAttr.addField("zyx", 5)
        INPUT_ATTR(eAttr)

        EulerScalarMath.inScalar = nAttr.create("scalar", "scalar", om2.MFnNumericData.kDouble, 0.0)
        INPUT_ATTR(nAttr)

        EulerScalarMath.inResRotOrder = eAttr.create("rotateOrderOutEuler", "rooe", 0)
        eAttr.addField("xyz", 0)
        eAttr.addField("yzx", 1)
        eAttr.addField("zxy", 2)
        eAttr.addField("xzy", 3)
        eAttr.addField("yxz", 4)
        eAttr.addField("zyx", 5)
        INPUT_ATTR(eAttr)

        outEulerX = uAttr.create("outEulerX", "oex", om2.MFnUnitAttribute.kAngle, 0.0)
        outEulerY = uAttr.create("outEulerY", "oey", om2.MFnUnitAttribute.kAngle, 0.0)
        outEulerZ = uAttr.create("outEulerZ", "oez", om2.MFnUnitAttribute.kAngle, 0.0)
        EulerScalarMath.outEuler = nAttr.create("outEuler", "oe", outEulerX, outEulerY, outEulerZ)
        OUTPUT_ATTR(nAttr)

        EulerScalarMath.addAttribute(EulerScalarMath.inOperation)
        EulerScalarMath.addAttribute(EulerScalarMath.inEuler)
        EulerScalarMath.addAttribute(EulerScalarMath.inEulerRotOrder)
        EulerScalarMath.addAttribute(EulerScalarMath.inScalar)
        EulerScalarMath.addAttribute(EulerScalarMath.inResRotOrder)
        EulerScalarMath.addAttribute(EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inOperation, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inEuler, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inEulerRotOrder, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inScalar, EulerScalarMath.outEuler)
        EulerScalarMath.attributeAffects(EulerScalarMath.inResRotOrder, EulerScalarMath.outEuler)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=no-self-use
        if plug != EulerScalarMath.outEuler:
            return om2.kUnknownParameter

        operation = dataBlock.inputValue(EulerScalarMath.inOperation).asShort()
        vEuler = dataBlock.inputValue(EulerScalarMath.inEuler).asVector()
        scalar = dataBlock.inputValue(EulerScalarMath.inScalar).asDouble()
        eulerRotOder = dataBlock.inputValue(EulerScalarMath.inEulerRotOrder).asShort()
        outRotOrder = dataBlock.inputValue(EulerScalarMath.inResRotOrder).asShort()

        eEuler = om2.MEulerRotation(vEuler, eulerRotOder)

        outEulerHandle = dataBlock.outputValue(EulerScalarMath.outEuler)

        if operation == 0:
            eEuler.reorderIt(outRotOrder)
            vResult = eEuler.asVector()
            outEulerHandle.setMVector(vResult)
        elif operation == 1:
            eEuler.reorderIt(outRotOrder)
            vScalar = om2.MVector(
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
            )
            vResult = eEuler.asVector() + vScalar
            outEulerHandle.setMVector(vResult)
        elif operation == 2:
            eEuler.reorderIt(outRotOrder)
            vScalar = om2.MVector(
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
                om2.MAngle(scalar, om2.MAngle.kDegrees).asRadians(),
            )
            vResult = eEuler.asVector() - vScalar
            outEulerHandle.setMVector(vResult)
        elif operation == 3:
            eEuler.reorderIt(outRotOrder)
            eOutEuler = eEuler * scalar
            vResult = eOutEuler.asVector()
            outEulerHandle.setMVector(vResult)

        outEulerHandle.setClean()

File: plug-ins/dev/cpp/source/n_gfUtilEulerScalarMath.cpp
#include "headers/n_gfUtilEulerScalarMath.h"


// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
EulerScalarMath::EulerScalarMath() {}

// Destructor.
EulerScalarMath::~EulerScalarMath() {}

MObject EulerScalarMath::inOperation;
MObject EulerScalarMath::inEuler;
MObject EulerScalarMath::inEulerX;
MObject EulerScalarMath::inEulerY;
MObject EulerScalarMath::inEulerZ;
MObject EulerScalarMath::inEulerRotOrder;
MObject EulerScalarMath::inScalar;
MObject EulerScalarMath::inResRotOrder;
MObject EulerScalarMath::outEuler;
MObject EulerScalarMath::outEulerX;
MObject EulerScalarMath::outEulerY;
MObject EulerScalarMath::outEulerZ;


void* EulerScalarMath::creator(){
    // Maya creator function.
    return new EulerScalarMath();
}

MStatus EulerScalarMath::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to TestNode class. Instances of TestNode will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnEnumAttribute eAttr;
    MFnUnitAttribute uAttr;
    MFnNumericAttribute nAttr;

    inOperation = eAttr.create("operation", "operation", 0, &status);
    eAttr.addField("No Operation", 0);
    eAttr.addField("Add", 1);
    eAttr.addField("Subtract", 2);
    eAttr.addField("Multiply", 3);
    INPUT_ATTR(eAttr);

    inEulerX = uAttr.create("eulerX", "ex", MFnUnitAttribute::kAngle, 0.0, &status);
    inEulerY = uAttr.create("eulerY", "ey", MFnUnitAttribute::kAngle, 0.0, &status);
    inEulerZ = uAttr.create("eulerZ", "ez", MFnUnitAttribute::kAngle, 0.0, &status);
    inEuler = nAttr.create("euler", "e", inEulerX, inEulerY, inEulerZ, &status);
    INPUT_ATTR(nAttr);

    inEulerRotOrder = eAttr.create("rotateOrderEuler", "roe", 0, &status);
    eAttr.addField("xyz", 0);
    eAttr.addField("yzx", 1);
    eAttr.addField("zxy", 2);
    eAttr.addField("xzy", 3);
    eAttr.addField("yxz", 4);
    eAttr.addField("zyx", 5);
    INPUT_ATTR(eAttr);

    inScalar = nAttr.create("scalar", "scalar", MFnNumericData::kDouble, 0.0, &status);
    INPUT_ATTR(nAttr);

    inResRotOrder = eAttr.create("rotateOrderOutEuler", "rooe", 0, &status);
    eAttr.addField("xyz", 0);
    eAttr.addField("yzx", 1);
    eAttr.addField("zxy", 2);
    eAttr.addField("xzy", 3);
    eAttr.addField("yxz", 4);
    eAttr.addField("zyx", 5);
    INPUT_ATTR(eAttr);

    outEulerX = uAttr.create("outEulerX", "oex", MFnUnitAttribute::kAngle, 0.0, &status);
    outEulerY = uAttr.create("outEulerY", "oey", MFnUnitAttribute::kAngle, 0.0, &status);
    outEulerZ = uAttr.create("outEulerZ", "oez", MFnUnitAttribute::kAngle, 0.0, &status);
    outEuler = nAttr.create("outEuler", "oe", outEulerX, outEulerY, outEulerZ, &status);
    OUTPUT_ATTR(nAttr);

    addAttribute(inOperation);
    addAttribute(inEuler);
    addAttribute(inEulerRotOrder);
    addAttribute(inScalar);
    addAttribute(inResRotOrder);
    addAttribute(outEuler);
    attributeAffects(inOperation, outEuler);
    attributeAffects(inEuler, outEuler);
    attributeAffects(inEulerRotOrder, outEuler);
    attributeAffects(inScalar, outEuler);
    attributeAffects(inResRotOrder, outEuler);

    return status;
}

MStatus EulerScalarMath::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    if ((plug != outEuler) &&
        (plug != outEulerX) &&
        (plug != outEulerY) &&
        (plug != outEulerZ)){
        return MStatus::kUnknownParameter;
    }

    short operation = dataBlock.inputValue(inOperation).asShort();
    double3 &euler = dataBlock.inputValue(inEuler).asDouble3();
    double scalar = dataBlock.inputValue(inScalar).asDouble();
    short eulerRotOrder = dataBlock.inputValue(inEulerRotOrder).asShort();
    short outRotOrder = dataBlock.inputValue(inResRotOrder).asShort();

    MEulerRotation eEuler = MEulerRotation(
        euler[0], 
        euler[1],
        euler[2],
        (MEulerRotation::RotationOrder) eulerRotOrder
    );

    MDataHandle outEulerHdle = dataBlock.outputValue(outEuler);
    MEulerRotation eScalar, eOutEuler;

    switch (operation)
    {
    case 0:
        eEuler.reorderIt((MEulerRotation::RotationOrder) outRotOrder);
        outEulerHdle.set3Double(eEuler.x, eEuler.y, eEuler.z);
        break;
    case 1:
        eEuler.reorderIt((MEulerRotation::RotationOrder) outRotOrder);
        eScalar = MEulerRotation(
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            (MEulerRotation::RotationOrder) outRotOrder
        );
        eOutEuler = eEuler + eScalar;
        outEulerHdle.set3Double(eOutEuler.x, eOutEuler.y, eOutEuler.z);
        break;
    case 2:
        eEuler.reorderIt((MEulerRotation::RotationOrder) outRotOrder);
        eScalar = MEulerRotation(
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            MAngle(scalar, MAngle::kDegrees).asRadians(),
            (MEulerRotation::RotationOrder) outRotOrder
        );
        eOutEuler = eEuler - eScalar;
        outEulerHdle.set3Double(eOutEuler.x, eOutEuler.y, eOutEuler.z);
        break;
    case 3:
        eEuler.reorderIt((MEulerRotation::RotationOrder) outRotOrder);
        eOutEuler = eEuler * scalar;
        outEulerHdle.set3Double(eOutEuler.x, eOutEuler.y, eOutEuler.z);
        break;
    }

    outEulerHdle.setClean();

    return MStatus::kSuccess;
}